packages/lib-classifier/src/store/AnnotatedSteps/AnnotatedSteps.js

```diff
@@ -113,16 +113,18 @@ const AnnotatedSteps = types.model('AnnotatedSteps', {
   function finish(){
     const { annotations, step } = self.latest
     undoManager.withoutUndo(() => {
+      step.validateTasks(annotations)
       step.completeTasks(annotations)
       _clearRedo()
     })
   }
   /** Redo the next step, or add a new step to history if there is no redo. */
   function next() {
-    const { nextStepKey } = self.latest
+    const { annotations, nextStepKey, step } = self.latest
     if (undoManager.canRedo) {
       _redo(nextStepKey)
     } else {
+      step.validateTasks(annotations)
```

Comment:
This may trigger a similar bug to #2200. On redo, changes to the transcription task won’t be validated. Instead, the original `validateTasks` action will be applied again.

Reply:
I've got a test workflow set-up with transcription lines followed by a drawing task.
https://localhost:8080/?env=staging&project=908&workflow=3372

I don't see a regression of #2200 on this branch 👍 

However, task validation is broken when you try to undo `step.validateTasks(annotations)` then redo it. I created an invalid transcription line, then pressed Next. The invalid line was removed but when I pressed Back, undo threw errors in the console (`Uncaught TypeError: "newSnapshot" is read-only`) and I'm still seeing the Back button rendered for the first task. I think that indicates that the UndoManager has broken. I think `step.validateTasks(annotations)` generates snapshots that can't be replayed by undo/redo.

Here's what I'm seeing for the task navigation on the first step when I go back from the question task to the drawing task. There's a Back button, which doesn't work.
![Screenshot of the task area showing a Back button onthe first task.](https://user-images.githubusercontent.com/59547/119480214-574f2900-bd49-11eb-9686-fbe4982d5131.png)

In #2199 I found that the solution to undo/redo bugs for `step.completeTasks` was to move that action into the `onClick` handler in the `NextButtonConnector`, so that we're not trying to record and replay snapshots for the step actions.

Reply:
Yes, I was aware of that PR and it's the reason I've held of merging this. In the future, if you submit a review, please actually use a status (i.e. request changes) if a bug is present so that the PR doesn't accidentally get merged with an approval prematurely and the status of the PR is easily seen in the PR list.